In [1]:
# Importing necessary libraries
import pandas as pd
import numpy as np

In [2]:
# Reading ExcelFile.
xls = pd.ExcelFile('Data.xlsx')

# Previewing sheet names inside excel file.
xls.sheet_names

['RH K-4 MOY Data',
 'MOY Export',
 'Syracuse',
 'Howard',
 'Swarthmore',
 'Spelman',
 'Williams',
 'Brooklyn',
 'Tuskegee',
 'Northwestern',
 'Georgetown',
 'Wisconsin']

In [3]:
# Combining all student records into a single Data Frame (df), excluding first two summary sheets
df = pd.concat(pd.read_excel(xls, sheet_name=xls.sheet_names[2:]), ignore_index=True)
print(df.shape)
df.head(25)

(274, 17)


Student_ Number   Grade_Level  Home_room  BOY Indep Level  \
0        236563003.0           0.0   Syracuse              AA   
1        238194344.0           0.0   Syracuse              AA   
2        239980246.0           0.0   Syracuse              AA   
3        240845834.0           0.0   Syracuse              AA   
4        239980196.0           0.0   Syracuse              AA   
5        240414854.0           0.0   Syracuse              AA   
6        240158048.0           0.0   Syracuse              AA   
7        240060129.0           0.0   Syracuse              AA   
8        240267807.0           0.0   Syracuse              AA   
9        236665122.0           0.0   Syracuse              AA   
10       239980253.0           0.0   Syracuse              AA   
11       240414904.0           0.0   Syracuse              AA   
12       238742233.0           0.0   Syracuse              AA   
13       239352479.0           0.0   Syracuse              AA   
14       239980238.0           0.0   Syracuse              AA   
15       240060145.0           0.0   Syracuse              AA   
16       240060210.0           0.0   Syracuse              AA   
17       239980279.0           0.0   Syracuse              AA   
18       239980154.0           0.0   Syracuse              AA   
19       239427610.0           0.0   Syracuse              AA   
20       240426700.0           0.0   Syracuse              AA   
21       239983976.0           0.0   Syracuse               B   
22       240158022.0           0.0   Syracuse              AA   
23       239980287.0           0.0   Syracuse              AA   
24       240415372.0           0.0   Syracuse                   

     Grade Level Equiv  Inst Level  Grade Level Equiv.1 MOY Indep Level  \
0                -0.25         NaN                                   AA   
1                -0.25         NaN                                   AA   
2                -0.25         NaN                                   AA   
3                -0.25         NaN                                   AA   
4                -0.25         NaN                                   AA   
5                -0.25         NaN                                    A   
6                -0.25         NaN                                    A   
7                -0.25         NaN                                    A   
8                -0.25         NaN                                    A   
9                -0.25         NaN                                    A   
10               -0.25         NaN                                    A   
11               -0.25         NaN                                    B   
12               -0.25         NaN                                    B   
13               -0.25         NaN                                    B   
14               -0.25         NaN                                    B   
15               -0.25         NaN                                    D   
16               -0.25         NaN                                    B   
17               -0.25         NaN                                    C   
18               -0.25         NaN                                    C   
19               -0.25         NaN                                    C   
20               -0.25         NaN                                    D   
21                0.33         NaN                                    D   
22               -0.25         NaN                                    D   
23               -0.25         NaN                                    F   
24                 NaN                                               AA   

    Grade Level Equiv.2  Inst Level.1  Grade Level Equiv.3   Accuracy (%)  \
0                 -0.25             A                  0.0           88.0   
1                 -0.25             A                  0.0           97.0   
2                 -0.25             A                  0.0           33.0   
3                                  AA                -

In [4]:
# previewing df
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 274 entries, 0 to 273
Data columns (total 17 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Student_ Number       274 non-null    float64
 1   Grade_Level           274 non-null    float64
 2   Home_room             274 non-null    object 
 3   BOY Indep Level       273 non-null    object 
 4    Grade Level Equiv    259 non-null    float64
 5    Inst Level           187 non-null    object 
 6    Grade Level Equiv.1  274 non-null    object 
 7   MOY Indep Level       271 non-null    object 
 8    Grade Level Equiv.2  274 non-null    object 
 9    Inst Level.1         264 non-null    object 
 10   Grade Level Equiv.3  254 non-null    object 
 11   Accuracy (%)         218 non-null    float64
 12   Fluency (0-3)        232 non-null    float64
 13   Comp Within (0-3)    251 non-null    float64
 14   Comp Beyond (0-3)    249 non-null    float64
 15   Comp About           1

A couple of noticings:

* The column names are inconsistently formatted and need to be cleaned.
* Some numerical columns need to be casted as such.
* Since the Non-null counts are inconsistent, further investigation is needed.

In [5]:
# Removing spaces and special characters from column names 
df.columns = df.columns.str.strip()
df.columns = df.columns.str.replace('_',' ')
    
# Renaming Student Number & Homeroom
df.insert(1,'OSIS', df['Student  Number'])
df.insert(2,'Homeroom', df['Home room'])

# Creating list of columns to drop
drop_cols = ['Home room','Student  Number']

# Dropping columns
df.drop(drop_cols, axis=1, inplace=True)

In [6]:
# Renaming columns for easier readibility
df.insert(3,'BOY IP Reading Level', df['BOY Indep Level'])
df.insert(4,'BOY IP Reading Grade Equiv', df['Grade Level Equiv'])
df.insert(5,'BOY Instr Reading Level', df['Inst Level'])
df.insert(6,'BOY Instr Reading Grade Equiv', df['Grade Level Equiv.1'])
df.insert(7,'MOY IP Reading Level', df['MOY Indep Level'])
df.insert(8,'MOY IP Reading Grade Equiv', df['Grade Level Equiv.2'])
df.insert(9,'MOY Instr Reading Level', df['Inst Level.1'])
df.insert(10,'MOY Instr Reading Grade Equiv', df['Grade Level Equiv.3'])

In [7]:
# Dropping columns
drop_cols1 = ['BOY Indep Level','Grade Level Equiv','Inst Level','Grade Level Equiv.1','MOY Indep Level',
            'Grade Level Equiv.2','Inst Level.1','Grade Level Equiv.3']

df.drop(drop_cols1, axis=1, inplace=True)

The reading levels are documented in capital letters. As a precaution, I will make all independent and instructional levels capital.

In [8]:
# Capitalizing all reading levels
reading_level_cols =  ['BOY IP Reading Level','BOY Instr Reading Level',
                       'MOY IP Reading Level', 'MOY Instr Reading Level']

for col in reading_level_cols:
    df[col]=df[col].str.upper()

for i, row in df.iterrows():
    print(f"Index: {i}, {row['BOY IP Reading Level']}, {row['MOY IP Reading Level']}")

Index: 0, AA, AA
Index: 1, AA, AA
Index: 2, AA, AA
Index: 3, AA, AA
Index: 4, AA, AA
Index: 5, AA, A
Index: 6, AA, A
Index: 7, AA, A
Index: 8, AA, A
Index: 9, AA, A
Index: 10, AA, A
Index: 11, AA, B
Index: 12, AA, B
Index: 13, AA, B
Index: 14, AA, B
Index: 15, AA, D
Index: 16, AA, B
Index: 17, AA, C
Index: 18, AA, C
Index: 19, AA, C
Index: 20, AA, D
Index: 21, B, D
Index: 22, AA, D
Index: 23, AA, F
Index: 24,  , AA
Index: 25, AA, B
Index: 26,  , AA
Index: 27, AA, AA
Index: 28, AA, AA
Index: 29, AA,  
Index: 30, AA, AA
Index: 31, AA,  
Index: 32, AA, A
Index: 33, AA, A
Index: 34, AA, A
Index: 35, AA, A
Index: 36, AA,  
Index: 37, AA,  
Index: 38, AA,  
Index: 39, AA, A
Index: 40, AA, B
Index: 41, AA, B
Index: 42, AA, B
Index: 43, AA, C
Index: 44, AA, C
Index: 45, AA, C
Index: 46, AA, C
Index: 47, AA, C
Index: 48, AA, D
Index: 49, AA, D
Index: 50, AA, H
Index: 51, AA, H
Index: 52, AA, nan
Index: 53,  ,  
Index: 54, A, AA
Index: 55, AA, A
Index: 56, B, B
Index: 57, B, B
Index: 58, AA, B
I

In [10]:
df.loc[24,'BOY IP Reading Level'] = np.nan

In [13]:
for i, row in df.iterrows():
    print(f"Index: {i}, {row['BOY IP Reading Level']}, {row['MOY IP Reading Level']}")
    i += 1

Index: 0, AA, AA
Index: 1, AA, AA
Index: 2, AA, AA
Index: 3, AA, AA
Index: 4, AA, AA
Index: 5, AA, A
Index: 6, AA, A
Index: 7, AA, A
Index: 8, AA, A
Index: 9, AA, A
Index: 10, AA, A
Index: 11, AA, B
Index: 12, AA, B
Index: 13, AA, B
Index: 14, AA, B
Index: 15, AA, D
Index: 16, AA, B
Index: 17, AA, C
Index: 18, AA, C
Index: 19, AA, C
Index: 20, AA, D
Index: 21, B, D
Index: 22, AA, D
Index: 23, AA, F
Index: 24, nan, AA
Index: 25, AA, B
Index: 26,  , AA
Index: 27, AA, AA
Index: 28, AA, AA
Index: 29, AA,  
Index: 30, AA, AA
Index: 31, AA,  
Index: 32, AA, A
Index: 33, AA, A
Index: 34, AA, A
Index: 35, AA, A
Index: 36, AA,  
Index: 37, AA,  
Index: 38, AA,  
Index: 39, AA, A
Index: 40, AA, B
Index: 41, AA, B
Index: 42, AA, B
Index: 43, AA, C
Index: 44, AA, C
Index: 45, AA, C
Index: 46, AA, C
Index: 47, AA, C
Index: 48, AA, D
Index: 49, AA, D
Index: 50, AA, H
Index: 51, AA, H
Index: 52, AA, nan
Index: 53,  ,  
Index: 54, A, AA
Index: 55, AA, A
Index: 56, B, B
Index: 57, B, B
Index: 58, AA, B

In [17]:
for col in df.columns:
    df[col] = df[col].str.replace(" ","")

AttributeError: Can only use .str accessor with string values!

In [15]:
df['BOY IP Reading Level'].str.replace(" ", "")

0      AA
1      AA
2      AA
3      AA
4      AA
       ..
269     Q
270     Q
271     R
272     R
273     R
Name: BOY IP Reading Level, Length: 274, dtype: object

In [9]:
# Removing potential spaces from all records

# Placing r in front creates a raw-string literal.
# Character class, \s, matches whitespace character (includes tabs and line breaks)
# Quantifier, *, matches a string that has a space followed by 0 or more spaces.
# Anchors, ^ and $, starts and ends with a space.

for col in df.select_dtypes(include=[object]):
    df[col] = df[col].replace(r'^\s*$', np.nan, regex=True)

NameError: name 'd' is not defined

In [ ]:
# Checking for completion 
df.info()

In [ ]:
# Getting summary of missing values
df.isna().sum()

A significant amount of data is missing from numerous columns. Rather than automatically drop these columns, I will use the key provided with this analysis to fill this missing data.

In [ ]:
# importing BOY info
sheet_url = 'https://docs.google.com/spreadsheets/d/1ctQdLgknVDQ5BPCS-O6UihzK5kPbbVwI-svT3cWKWfw/edit#gid=1026408630'
url1 = sheet_url.replace('/edit#gid=', '/export?format=csv&gid=')        
df_BOY_levels = pd.read_csv(url1, header=1)
df_BOY_levels.head(27)

In [ ]:
# Importing MOY info
sheet_url = 'https://docs.google.com/spreadsheets/d/1ctQdLgknVDQ5BPCS-O6UihzK5kPbbVwI-svT3cWKWfw/edit#gid=1314641425'
url2 = sheet_url.replace('/edit#gid=', '/export?format=csv&gid=')        
df_MOY_levels = pd.read_csv(url2, header=1)
df_MOY_levels.head(27)

BOY and MOY Independent Reading Levels & Grade Equivalents are identical.  I will create a dictionary with these level to assist with missing values.

In [ ]:
# Dictionary 
alpha_keys = list(df_BOY_levels['Independent Reading Level'].unique())
alpha_values = list(df_BOY_levels['Grade Equivalent'].unique())

levels = dict(zip(alpha_keys, alpha_values))
levels

In [ ]:
# Locating records with missing BOY IP level data
df.loc[(df['BOY IP Level'].isna())]

This scholar is Swarthmore has data for the MOY assesment, but not the BOY assessment. I will assume this means the student was not assessed in the BOY. Since it is the beginning I will also assume this means that the student was not present at the time of assessment. I will replace the BOY records with NT (Not Tested).

In [ ]:
# Addressing missing data for student 238231914
df['BOY IP Level'].fillna('NT', inplace=True)

df.loc[(df['BOY IP Level']=='NT'), 'BOY IP Grade Equiv']='NT'
df.loc[(df['BOY IP Level']=='NT'), 'BOY Instr Level']='NT'
df.loc[(df['BOY IP Level']=='NT'), 'BOY Instr Grade Equiv']='NT'

In [ ]:
# Verifying accuracy
df.loc[(df['BOY IP Level'].isna())]

In [ ]:
# Locating records with missing BOY Grade Equivalent data
print(df.loc[df['BOY IP Grade Equiv'].isna()].shape)
df.loc[df['BOY IP Grade Equiv'].isna()]

Upon further investigation, only one BOY IP Level was included (student with OSIS # 203306691). I will use the key to fill this missing BOY Grade Equiv for this student. For the other students, my assumption is that a majority do not have beginning year data because they are no longer enrolled in the school. I assumed this because their MOY data is also missing. I will fill those values with NP and eventually drop them from the data frame.

In [ ]:
# Replacing grade equivalence for level H
df.loc[(df['OSIS']==203306691),'BOY IP Grade Equiv']=1.67

# Checking to see if change was made.
df.loc[df['OSIS']==203306691]

In [ ]:
# Locating records with missing BOY Inst Level data
df.loc[df['BOY Instr Level'].isna()]

In [ ]:
# Filling Instructional Level given BOY Independent Levels
i = 0
for i, row in df.iterrows():
    for i in range(len(alpha_keys)-1):
        if pd.isnull(row['BOY Instr Level']):
            for level in alpha_keys:
                df.loc[(df['BOY IP Level']==alpha_keys[i]), 'BOY Instr Level']=alpha_keys[i+1]

In [ ]:
# Checking for accuracy 
df.loc[df['BOY Instr Level'].isna()]

In [ ]:
# Spot-checking with previous student
df.loc[df['OSIS']==203306691]

In [ ]:
# Addressing BOY Instructional Level since all were missing, no conditionals were used.
i = 0
for level, equiv in levels.items():
    df.loc[(df['BOY Instr Level']==level), 'BOY Instr Grade Equiv'] = equiv
    
# Checking for accuracy    
df.tail(25)

In [ ]:
# Verifying there are no records with missing BOY Instr Grade Equiv
df.loc[df['BOY Instr Level'].isna()]

In [ ]:
# Completing cleaning process for MOY Data

In [ ]:
# Locating records with missing MOY Indepent Level data
df.loc[df['MOY IP Level'].isna()]

There are 3 students with missing MOY IP Level, but there is an instructional level. The instructional level for the BOY was one level above the student's IP BOY level. I will assume MOY Independent Levels follow a similar pattern. MOP IP Levels will be one level below MOY instructional levels.

In [ ]:
# Filling MOY Independent Levels given MOY Instructional Level
i = 1
for i in range(len(alpha_keys)-1):
    if (df['MOY IP Level'].isna().any()):
        for level in alpha_keys:
            df.loc[(df['MOY Instr Level']==alpha_keys[i]), 'MOY IP Level']=alpha_keys[i-1]
            i=+1

In [ ]:
# Spot-checking for accuracy    
# df.loc[df['OSIS']==234033942]
df.loc[df['OSIS']==203306691]

In [ ]:
df['MOY IP Level'].notnull().shape

We now have 274 complete MOY Independent Levels. I will move on to investigate other columns.

In [ ]:
# Locating records with missing MOY data
print(df.loc[df['MOY IP Grade Equiv'].isna()].shape)

In [ ]:
df.loc[df['MOY IP Grade Equiv'].isna()]

In [ ]:
# Addressing MOY IP Grade Equiv 
for level, equiv in levels.items():
    df.loc[(df['MOY IP Level']==level), 'MOY IP Grade Equiv'] = equiv
    
# Checking for accuracy    
df.loc[df['MOY IP Grade Equiv'].isna()].shape

In [ ]:
# Next column: Instructional Level.
df.loc[df['MOY Instr Level'].isna()].shape

In [ ]:
# Investigating 10 missing records.
df.loc[df['MOY Instr Level'].isna()]

In [ ]:
# Filling MOY Instructional Level given MOY Independent Levels
i = 0
for i in range(len(alpha_keys)-1):
    for level in alpha_keys:
        df.loc[(df['MOY IP Level']==alpha_keys[i]), 'MOY Instr Level']=alpha_keys[i+1]
        i=+1

# Checking for completeness
df.loc[df['MOY Instr Level'].isna()]

In [ ]:
# Filling MOY Instr Grade Equiv given Instructional Levels
i = 0
for level, equiv in levels.items():
    df.loc[(df['MOY Instr Level']==level), 'MOY Instr Grade Equiv'] = equiv
    
# Checking for completeness
df.loc[df['MOY Instr Grade Equiv'].isna()]

The data shows that students missing MOY data also were students who are also missing in the BOY. My assumption is that these students were registered, but didnot attend the school. I will drop these records from the data frame.

In [ ]:
df.dropna(subset=['MOY Instr Grade Equiv'], inplace=True)

In [ ]:
df['MOY Instr Grade Equiv'].isna().sum()

In [ ]:
df.info()

In [ ]:
# Investigating Growth

In [ ]:
for i, row in df.iterrows():
    print(f"Index: {i}, {row['BOY IP Level']}, {row['MOY IP Level']}")

The 4th and 5th student in the data frame shows a BOY Independent Level of AA and MOY Independent Level of Z. A student growing that much in less than a year is highly unlikely. An investigation into these outliers are required.

In [ ]:
# Changing data to numerical type
df[['BOY IP Grade Equiv','BOY Instr Grade Equiv','MOY IP Grade Equiv','MOY Instr Grade Equiv']
  ] = df[['BOY IP Grade Equiv','BOY Instr Grade Equiv','MOY IP Grade Equiv','MOY Instr Grade Equiv']
        ].apply(pd.to_numeric, errors='coerce')

In [ ]:
# Investigating Outliers by creating a column showing the difference between BOY and MOY levels.
df.isna().sum()
df['BOY-MOY Growth'] = abs(df['MOY IP Grade Equiv'] - df['BOY IP Grade Equiv'])
df['BOY-MOY Growth']

In [ ]:
df['BOY-MOY Growth'].plot(title="Student Growth");

Since this analysis compares student's reading levels from the beginning of the school to the middle of school year, I expect to see a few cases of regression, no growth, and atleast a half a year of growth in reading levels. According to F&P Instructional Level Expectations for Reading, students can grow 3-4 levels within a school year. For example, a first grader is expected to enter the grade reading on level E or above and be on level K or above by the end of the school year. For this part of the analysis, I will isolate students who grew more than two grade levels and investigate their growth.

In [ ]:
# Creating column with on grade level, above grade level, below grade level.
df['BOY Ranking'] = np.nan
df['MOY Ranking'] = np.nan

In [ ]:
# Create column for on grade level or not for BOY and MOY.